feat: Add CataclysmicMap (Dungeon Map)

My-Name-Is-Jeff opened the pull request.

My-Name-Is-Jeff committed:
feat: render map loaded rooms on CataclysmicMap

My-Name-Is-Jeff committed:
fix: properly rotate CataclysmicMap based on GuiElement data

My-Name-Is-Jeff committed:
fix: mapHideInBoss has no effect

My-Name-Is-Jeff committed:
feat: add highlight wither doors

My-Name-Is-Jeff committed:
fix: Map heads don't render if party size isn't 5

My-Name-Is-Jeff committed:
feat: Use loaded entity locations for Dungeon Map

TakoTheDev commented on src/main/kotlin/gg/skytils/skytilsmod/features/impl/dungeons/cataclysmicmap/CataclysmicMap.kt:
```diff
@@ -0,0 +1,111 @@
+    @SubscribeEvent
+    fun onWorldLoad(event: WorldEvent.Load) {
+        DungeonInfo.reset()
+        MapUtils.calibrated = false
```
Why not just use the reset function?

My-Name-Is-Jeff committed:
feat: make it easier to open CataclysmicMap menu

My-Name-Is-Jeff committed:
fix: black border around map player heads

My-Name-Is-Jeff committed:
feat: add question mark icon for loaded rooms

My-Name-Is-Jeff committed:
refactor: rename RoomState.LOADED to UNOPENED

TakoTheDev commented on src/main/kotlin/gg/skytils/skytilsmod/features/impl/dungeons/cataclysmicmap/core/CataclysmicMapConfig.kt:
```diff
@@ -0,0 +1,411 @@
+    @Property(
+        name = "Center Map",
+        type = PropertyType.SWITCH,
+        description = "Centers the map on the player if Rotate Map is enabled.",
```
Should probably depend on Rotate map right?

TakoTheDev replied:
Actually you can probably only do one of these suggestions? idk

TakoTheDev commented on src/main/kotlin/gg/skytils/skytilsmod/features/impl/dungeons/cataclysmicmap/core/CataclysmicMapConfig.kt:
```diff
@@ -0,0 +1,411 @@
+    @Property(
+        name = "Dynamic Rotate",
+        type = PropertyType.SWITCH,
+        description = "Keeps the entrance room at the bottom. Does not work with rotate map.",
```
Should probably be in a selector list with rotate map?

TakoTheDev commented:
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/fb68052f-f23c-45b5-915a-aa2ea4b0a6a5)
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/f7504907-e621-4eb6-8528-a093a03bb2d7)
It seems one player is always missing and instead rendered on top of another player or smth?

And the names should probably not rotate with the head
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/6722d29f-4b73-47dd-9ec8-6f591ed4ea63)

Also while a door is opening it shows the no key highlight, possibly make it not show at all once you open it?

![image](https://github.com/Skytils/SkytilsMod/assets/74718793/acbf2653-52df-4fae-a2cd-a56348b3a969)
If you know how many secrets are done in a room, maybe show it like 3/4 or ?/4 if not known?

![image](https://github.com/Skytils/SkytilsMod/assets/74718793/affa692b-e282-4116-8853-760a2f6625fc)
Maybe show checkmark for yellow room?

![image](https://github.com/Skytils/SkytilsMod/assets/74718793/19c8c7aa-55f0-43a9-9493-e22bfba81d40)
room has no secrets but is done, maybe show some indicator for that in the secrets replace checkmark setting?

My-Name-Is-Jeff committed:
fix: CataclysmicMap assigns heads incorrectly

My-Name-Is-Jeff commented on src/main/kotlin/gg/skytils/skytilsmod/features/impl/dungeons/cataclysmicmap/core/DungeonMapPlayer.kt:
```diff
@@ -0,0 +1,46 @@
+
+    /** Has information from player entity been loaded */
+    var playerLoaded = false
+    // TODO: fix wrong icon count when players die
```
needs research

TakoTheDev commented:
aware that it is a temp commit, but still why not
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/1efb6ad0-deb4-4da8-9210-4b3b6d615f45)

My-Name-Is-Jeff committed:
chore: refactor CataclysmicMap

chore: remove unused fields

chore: refactor again

fix: redundant code

fix: add sba attribution

My-Name-Is-Jeff committed:
feat: Add Cataclysmic Map

chore: Add Bloom credit

remove: useless mimic room color

My-Name-Is-Jeff committed:
fix: dungeon map tries to get data from score summary

My-Name-Is-Jeff committed:
fix: Dungeon Map question mark is not drawn in edge cases


refactor: cleanup code

My-Name-Is-Jeff committed:
fix: stop updating dungeon map after entering boss room

fix: don't draw heads after entering boss

My-Name-Is-Jeff committed:
fix: change Dungeon Map failed rooms' name text color to red

My-Name-Is-Jeff committed:
fix: change Cataclysmic Map rendering logic for 0 secrets

fix: inverted boolean condition on checkmark drawing

fix: rebase issue

TakoTheDev commented:
Looks weird
---
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/b0481972-c1cb-4d47-9242-2dcffe310f77)

Looks weird
---
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/96f49f3d-c55e-461c-87a6-d1ec75401c2a)

Dynamic Rotate when the Entrance is at the top (likely happens when it is to the side as well):
---
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/843fccb3-d6af-44dc-a272-e7e15753dd23)

> Also while no screenshot of it rn, I still don't think the names should rotate with the head.

Oh also, if you're out of render distance, while someone is blood rushing, the door highlights stay until you load the chunks(?) in, which is not that bad of an issue but exists:
---
![image](https://github.com/Skytils/SkytilsMod/assets/74718793/c744574c-42fa-4b48-9364-d4c2e040f471)

My-Name-Is-Jeff committed:
fix: DungeonListener#team is incorrect when rejoining dungeon


fixes NPE
fixes issue where tab list party is "loaded"

My-Name-Is-Jeff committed:
feat: Update Dungeon Door opened state from Dungeon Map

Door opened state isn't updated when chunks aren't loaded

My-Name-Is-Jeff committed:
refactor: Change DungeonMapParser to a singleton

fix: clear DungeonMapColorParser on calibration

feat: cache last known Dungeon Map

updates dungeon map even when not in inventory

requires a bug fix from Hypixel Alpha to work properly

My-Name-Is-Jeff committed:
feat: add CataclysmicMap to LegalGui
